feat(Stream): support multiple streams #40

OCannings opened the pull request:
A new route has been added: `/stream/:streamId` which allows us to
connect to a stream-specific topic.

The socket middleware has been updated to support switching topics when
ever a new `socket_id` is used within an action payload.

TODO: write some tests.

OCannings committed:
feat(Stream): support multiple streams

A new route has been added: `/stream/:streamId` which allows us to
connect to a stream-specific topic.

The socket middleware has been updated to support switching topics when
ever a new `socket_id` is used within an action payload.

TODO: write some tests.

OCannings commented:
https://github.com/ItsJuice/juicebox/issues/40

ljkbennett commented:
https://github.com/ItsJuice/juicebox/projects/1#card-394152

OCannings committed:
feat(Stream): update queue on connect

The socket middleware has been refactored to support a dedicated
CONNECT_TO_STREAM action.

All socket events now use the generic "remote.action" event name, this
allows us to have a catch-all "rebroadcaster" in the socket middleware.

ljkbennett commented on web/channels/stream_channel.ex:
```diff
@@ -8,7 +8,11 @@ defmodule Juicebox.StreamChannel do
+
+    send(self, {:after_join, stream_id})
+
     {:ok, %{ queue: queue }, socket}
```
You are doing a push down the socket with the queue after join, but doesn't this line do that?

ljkbennett replied:
Ah wait, I saw you removed the connect action on the client. We don't need to send the queue here in that case, just return `{:ok, socket}`

ljkbennett commented on web/channels/stream_channel.ex:
```diff
@@ -21,15 +25,23 @@ defmodule Juicebox.StreamChannel do
     IO.inspect "Queue updated"
 
-    push socket, "queue.updated", payload
+    push socket, "remote.action", %{ payload | type: "QUEUE_UPDATED" }
```
Could we use the `QUEUE_UPDATED` on the server too, instead of `"queue.updated"`? Then the default could just be to forward blind, something like:

``` elixir
def handle_out(action_type, payload, socket) do
  push socket, "remote.action", %{ payload | type: action_type }
  {:noreply, socket}
end
```

If we needed to override it then we could just intercept them here.

OCannings committed:
refactor(Sockets): create sockets index file

A `sockets/index.js` file has been created to centralise all of the
socket related exports.

The "route subscriber" which is responsible for automatically connecting
to the stream's channel when the route changes has been moved into its
own file `sockets/subscribe.js`

OCannings committed:
test(Sockets): fill in missing socket tests

For sockets/actions and sockets/subscribe. Also update middleware tests
to match latest implementation.

OCannings committed:
refactor(Channel): support generic events

`handle_info/2` has been updated to support the forwarding of arbitrary
actions to the client. As long as the event Map has a `type` attribute
it will be sent to all connected clients.

`handle_out` has been removed from the channel - it appears to be unused.
